backend/optimize.ml

```diff
@@ -0,0 +1,210 @@
+        if !Flambda_backend_flags.cfg_peephole_optimize_track
+        then update_csv "remove_useless_mov";
+        DLL.delete_curr snd;
+        Some fst)
+      else None
+    | _ -> None)
+  | _ -> None
+
+(* Logical condition for simplifying the following case: 
+  {v 
+    <op 1> const1, r 
+    <op 2> const2, r 
+  v}
+
+  to: 
+  {v 
+    <op 1> (const1 <op 2> const2), r 
+  v}
+
+   Where <op 1> and <op 2> can be any two binary operators that are associative
+   and const1 and const2 are immediate values. *)
+
+(* CR-soon gtulba-lecu for gtulba-lecu: implement the rest of Intop_imm cases 
+   (only Imulh left to optimize if possible)
+    
+    it may be the case that after Iadd, Isub folding the result is 0, treat this case.
+*)
+let are_compatible op1 op2 imm1 imm2 =
+  match (op1: Mach.integer_operation), (op2: Mach.integer_operation) with
+  | Mach.Iand, Mach.Iand -> Some (Mach.Iand, imm1 land imm2)
+  | Ior, Ior -> Some (Mach.Ior, imm1 lor imm2)
+  | Ixor, Ixor -> Some (Mach.Ixor, imm1 lxor imm2)
+  | Ilsl, Ilsl -> Some (Mach.Ilsl, min (imm1 + imm2) Sys.int_size)
+  | Ilsr, Ilsr -> Some (Mach.Ilsr, min (imm1 + imm2) Sys.int_size)
+  | Iasr, Iasr -> Some (Mach.Iasr, min (imm1 + imm2) Sys.int_size)
+  | Iadd, Iadd -> 
+    if Misc.no_overflow_add imm1 imm2 then
+      Some (Mach.Iadd, imm1 + imm2)
+    else None
+  | Iadd, Isub -> 
+    if imm1 >= imm2 then
+      Some (Mach.Iadd, imm1 - imm2)
+    else Some (Mach.Isub, imm2 - imm1)
+  | Isub, Isub -> 
+    if Misc.no_overflow_add imm1 imm2 then
+      Some (Mach.Iadd, imm1 + imm2)
+    else None
+  | Isub, Iadd -> 
+    if imm1 >= imm2 then
+      Some (Mach.Isub, imm1 - imm2)
+    else Some (Mach.Iadd, imm2 - imm1)
+  | Imul, Imul -> 
+    if Misc.no_overflow_mul imm1 imm2 then
+      Some (Mach.Imul, imm1 * imm2)
+    else None    
+  | Idiv, Idiv -> 
+    if Misc.no_overflow_mul imm1 imm2 then
+      Some (Mach.Idiv, imm1 * imm2)
+    else None
+  | Imod, Imod ->
+    if (imm1 mod imm2) = 0 then
+      Some (Mach.Imod, imm2)
+    else None
+  | _ -> None
+
+let fold_intop_imm (fst : Cfg.basic Cfg.instruction DLL.cell)
+    (snd : Cfg.basic Cfg.instruction DLL.cell) =
+  let fst_val = DLL.value fst in
+  let snd_val = DLL.value snd in
+  if are_equal_regs fst_val.arg.(0) snd_val.arg.(0)
+  then
+    match fst_val.desc, snd_val.desc with
+    | Op (Intop_imm (op1, imm1)), Op (Intop_imm (op2, imm2)) ->
+      (match are_compatible op1 op2 imm1 imm2 with
+      | Some (op, imm) -> 
+        (if !Flambda_backend_flags.cfg_peephole_optimize_track
+        then update_csv "fold_intop_imm";
+        DLL.insert_before fst
+          { fst_val with desc = Cfg.Op (Intop_imm (op, imm)) };
+        let new_cell = DLL.prev fst in
+        DLL.delete_curr fst;
+        DLL.delete_curr snd;
+        new_cell)
+      | _ -> None)
+    | _ -> None
+  else None
+
+(* CR gtulba-lecu for xclerc: I am thinking of having this functionality in the
+   dll. It feels useful, what do you think? *)
```

Comment:
The operation might indeed be useful.

However, in the present case I think it
makes the `optimize_body'` function
harder to follow. It might be simpler
to have the "rewrite" functions return
the cell from which to continue the
iteration.